let backends set info on the lowered_backend.meta from preprocess

JacobSzwejbka opened the pull request:
Summary: We want some delegates to be able to test experimental features that interact with the core runtime. Typically when doing this there is an expectation of tight coupling in the full lowering process so we arent super concerned about UX. The backend api is pretty hard to change because the code is mostly owned by 3rd parties, so we want to be really sure when we change it that its correct. By exposing this module.meta hookup we can test things in a controled maner without changing the blessed api before we are ready.

Differential Revision: D81466391

digantdesai commented on exir/backend/backend_details.py:
```diff
@@ -32,6 +32,9 @@ class PreprocessResult:
     data_store_output: Optional[NamedDataStoreOutput] = None
 
+    # Optional delegate-specific information that will be added to the lowered_module's metadata.
+    delegate_info: Optional[Any] = None
```
Seems a little unstructured, but OK I guess. How about `delegate_info_meta` just to discourage people to put flatbuffers in here :p

hsharma35 replied:
Doing anything with delegated blob means we need to disassemble - modify - assemble the blob again. This is not expected to be trivial for backends.

JacobSzwejbka replied:
>How is it different than having it as part of the delegated blob?  

Some delegate blobs are not trivial to unpack is the main motivation. AOTI for instance would just be compiled code iiuc.

mergennachin replied:
@JacobSzwejbka 

This is a meta information that won't be serialized as part of the PTE file, right? 

If that's the case, can you document it here? 

Also I'd like to rename delegate_info to be something like `_delegate_info_meta` or `_delegate_info_internal` or just `_meta`?

JacobSzwejbka replied:
yes it just gets added to the fxGraph. Updated the name and comment

JacobSzwejbka committed:
let backends set info on the lowered_backend.meta from preprocess (#13856)

Summary:

We want some delegates to be able to test experimental features that interact with the core runtime. Typically when doing this there is an expectation of tight coupling in the full lowering process so we arent super concerned about UX. The backend api is pretty hard to change because the code is mostly owned by 3rd parties, so we want to be really sure when we change it that its correct. By exposing this module.meta hookup we can test things in a controled maner without changing the blessed api before we are ready.

Reviewed By: mergennachin, hsharma35

Differential Revision: D81466391
